lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs

```diff
@@ -0,0 +1,228 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# LANGUAGE TypeFamilies #-}
+-- |
+-- Copyright: © 2021 IOHK
+-- License: Apache-2.0
+--
+-- Address books consist of a 'Prologue'
+-- and discovered addresses ('Discoveries').
+module Cardano.Wallet.DB.Sqlite.AddressBook
+    ( AddressBookIso (..)
+    , Prologue (..)
+    , Discoveries (..)
+    , SeqAddressList (..)
+    )
+  where
+
+import Prelude
+
+import Cardano.Address.Derivation
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , MkKeyFingerprint (..)
+    , NetworkDiscriminant (..)
+    , PaymentAddress (..)
+    , Role (..)
+    , SoftDerivation (..)
+    )
+import Cardano.Wallet.Primitive.AddressDerivation.SharedKey
+    ( SharedKey (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( GetPurpose )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..), AddressState (..) )
+import Data.Generics.Internal.VL
+    ( Iso', iso, withIso )
+import Data.Kind
+    ( Type )
+import Data.Map.Strict
+    ( Map )
+import Data.Proxy
+    ( Proxy (..) )
+import Data.Type.Equality
+    ( type (==) )
+import Data.Typeable
+    ( Typeable )
+
+import qualified Cardano.Wallet.Primitive.AddressDiscovery.Random as Rnd
+import qualified Cardano.Wallet.Primitive.AddressDiscovery.Sequential as Seq
+import qualified Cardano.Wallet.Primitive.AddressDiscovery.Shared as Shared
+import qualified Cardano.Wallet.Primitive.Types.Address as W
+import qualified Data.Map.Strict as Map
+
+{-------------------------------------------------------------------------------
+    AddressBook isomorphism
+-------------------------------------------------------------------------------}
+-- | FIXME LATER during ADP-1043:
+-- Move 'Prologue' and 'Discoveries' closer into address types.
+class AddressBookIso s where
+    -- | Address information contained in the prologue of the address book,
+    -- such as public keys or the address gap.
+    data Prologue s :: Type
+    -- | Addresses that were collected during discovery on the blockchain.
+    data Discoveries s :: Type
+
+    -- | Isomorphism between the address book type 's'
+    -- and its two components.
+    addressIso :: Iso' s (Prologue s, Discoveries s)
+
+{-------------------------------------------------------------------------------
+    Sequential address book
+-------------------------------------------------------------------------------}
+-- | Sequential list of addresses
+newtype SeqAddressList (c :: Role) = SeqAddressList [(W.Address, W.AddressState)]
```

Comment:
Why we have here list? Because we are implying that the sequence of addresses correspond to address indices?So we have first entry corresponding to soft index ix=0, etc.? Why not to have a map with soft index as key? As we plan to enlarge this list (via `(++) [new address]`) or rollback usually with operations in the end of it...it seems the list could be quite expensive - naively thinking. Maybe strict map would be better here and we would not loose the capabilities here? What do you think about this argument?

Reply:
I completely agree that we will need a more efficient data structure than lists here — which is why I started refactoring the address pool logic, and was able to implement it for `SharedStates` in the upcoming PR #3068 😄. Step by step.

For the present PR, the list is sufficient: At the moment, the list is created when loading from the database and then immediately converted into an address pool through the `addressIso`. Similar procedure when storing to the database. As long as the conversion with `addressIso` happens just at these points, immediately before or after the database is accessed, we do pretty much exactly the same thing that the previous version of our code did, so we do not pay a performance penalty for now. Of course, once the isomorphism is used in the context of `Checkpoints (Discoveries s, UTxO,…)`, we need it to be faster. But that is for another PR.

Reply:
ok. fair enough